// @heroui/react/dist/components/dropdown/index.js
"use strict";
import { DropdownRoot, DropdownSubmenuTrigger, DropdownSubmenuIndicator, DropdownItemIndicator, DropdownItem, DropdownSection, DropdownMenu, DropdownPopover, DropdownTrigger } from './dropdown.js';
export { dropdownVariants } from '@heroui/styles';

/* -------------------------------------------------------------------------------------------------
 * Compound Component
 * -----------------------------------------------------------------------------------------------*/
const Dropdown = Object.assign(DropdownRoot, {
  Root: DropdownRoot,
  Trigger: DropdownTrigger,
  Popover: DropdownPopover,
  Menu: DropdownMenu,
  Section: DropdownSection,
  Item: DropdownItem,
  ItemIndicator: DropdownItemIndicator,
  SubmenuIndicator: DropdownSubmenuIndicator,
  SubmenuTrigger: DropdownSubmenuTrigger
});

export { Dropdown, DropdownItem, DropdownItemIndicator, DropdownMenu, DropdownPopover, DropdownRoot, DropdownSection, DropdownSubmenuIndicator, DropdownSubmenuTrigger, DropdownTrigger };
