// @heroui/react/dist/components/table/index.js
"use strict";
import { TableRoot, TableSortableColumnHeader, TableScrollContainer, TableRow, TableResizableContainer, TableLoadMoreContent, TableLoadMoreItem, TableHeader, TableFooter, TableContent, TableColumnResizer, TableColumn, TableCollection, TableCell, TableBody } from './table.js';
export { tableVariants } from '@heroui/styles';

/* -------------------------------------------------------------------------------------------------
 * Compound Component
 * -----------------------------------------------------------------------------------------------*/
const Table = Object.assign(TableRoot, {
  Body: TableBody,
  Cell: TableCell,
  Collection: TableCollection,
  Column: TableColumn,
  ColumnResizer: TableColumnResizer,
  Content: TableContent,
  Footer: TableFooter,
  Header: TableHeader,
  LoadMore: TableLoadMoreItem,
  LoadMoreContent: TableLoadMoreContent,
  ResizableContainer: TableResizableContainer,
  Root: TableRoot,
  Row: TableRow,
  ScrollContainer: TableScrollContainer,
  SortableColumnHeader: TableSortableColumnHeader
});

export { Table, TableBody, TableCell, TableCollection, TableColumn, TableColumnResizer, TableContent, TableFooter, TableHeader, TableLoadMoreContent, TableLoadMoreItem, TableResizableContainer, TableRoot, TableRow, TableScrollContainer, TableSortableColumnHeader };
