// @heroui/react/dist/components/toast/index.js
"use strict";
import { Toast as Toast$1, ToastCloseButton, ToastActionButton, ToastDescription, ToastTitle, ToastIndicator, ToastContent, ToastProvider } from './toast.js';
import { toast, ToastQueue } from './toast-queue.js';
export { toastQueue } from './toast-queue.js';
export { toastVariants } from '@heroui/styles';
export { DEFAULT_GAP, DEFAULT_MAX_VISIBLE_TOAST, DEFAULT_TOAST_TIMEOUT } from './constants.js';

/* -------------------------------------------------------------------------------------------------
 * Compound Component
 * -----------------------------------------------------------------------------------------------*/
const Toast = Object.assign(Toast$1, {
  Provider: ToastProvider,
  Content: ToastContent,
  Indicator: ToastIndicator,
  Title: ToastTitle,
  Description: ToastDescription,
  ActionButton: ToastActionButton,
  CloseButton: ToastCloseButton,
  Queue: ToastQueue,
  toast
});

export { Toast, ToastActionButton, ToastCloseButton, ToastContent, ToastDescription, ToastIndicator, ToastProvider, ToastQueue, ToastTitle, toast };
