// @heroui/react/dist/components/tabs/tabs.js
"use client";
import { tabsVariants } from '@heroui/styles';
import React__default, { createContext, useContext } from 'react';
import { SelectionIndicator } from 'react-aria-components/SelectionIndicator';
import { Tabs, TabPanel as TabPanel$1, Tab as Tab$1, TabList as TabList$1 } from 'react-aria-components/Tabs';
import { composeTwRenderProps, composeSlotClassName } from '../../utils/compose.js';
import { dom } from '../../utils/dom.js';
import { jsx } from 'react/jsx-runtime';

const TabsContext = /*#__PURE__*/createContext({});

/* -------------------------------------------------------------------------------------------------
 * Tabs Root
 * -----------------------------------------------------------------------------------------------*/

const TabsRoot = ({
  children,
  className,
  orientation = "horizontal",
  variant,
  ...props
}) => {
  const slots = React__default.useMemo(() => tabsVariants({
    variant
  }), [variant]);
  return /*#__PURE__*/jsx(TabsContext, {
    value: {
      orientation,
      slots
    },
    children: /*#__PURE__*/jsx(Tabs, {
      ...props,
      className: composeTwRenderProps(className, slots.base()),
      "data-slot": "tabs",
      orientation: orientation,
      children: children
    })
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tabs List Container
 * -----------------------------------------------------------------------------------------------*/

const TabListContainer = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(TabsContext);
  return /*#__PURE__*/jsx(dom.div, {
    className: composeSlotClassName(slots?.tabListContainer, className),
    "data-slot": "tabs-list-container",
    ...props,
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tabs List
 * -----------------------------------------------------------------------------------------------*/

const TabList = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(TabsContext);
  return /*#__PURE__*/jsx(TabList$1, {
    ...props,
    className: composeTwRenderProps(className, slots?.tabList()),
    "data-slot": "tabs-list",
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tab
 * -----------------------------------------------------------------------------------------------*/

const Tab = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(TabsContext);
  return /*#__PURE__*/jsx(Tab$1, {
    ...props,
    className: composeTwRenderProps(className, slots?.tab()),
    "data-slot": "tabs-tab",
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tab Indicator
 * -----------------------------------------------------------------------------------------------*/

const TabIndicator = ({
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(TabsContext);
  return /*#__PURE__*/jsx(SelectionIndicator, {
    className: composeSlotClassName(slots?.tabIndicator, className),
    "data-slot": "tabs-indicator",
    ...props
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tab Panel
 * -----------------------------------------------------------------------------------------------*/

const TabPanel = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(TabsContext);
  return /*#__PURE__*/jsx(TabPanel$1, {
    ...props,
    className: composeTwRenderProps(className, slots?.tabPanel()),
    "data-slot": "tabs-panel",
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tab Separator
 * -----------------------------------------------------------------------------------------------*/

const TabSeparator = ({
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(TabsContext);
  return /*#__PURE__*/jsx(dom.span, {
    "aria-hidden": "true",
    className: composeSlotClassName(slots?.separator, className),
    "data-slot": "tabs-separator",
    ...props
  });
};

export { Tab, TabIndicator, TabList, TabListContainer, TabPanel, TabSeparator, TabsRoot };
