// @heroui/react/dist/components/tooltip/tooltip.js
"use client";
import { tooltipVariants } from '@heroui/styles';
import { mergeProps } from '@react-aria/utils';
import React__default, { createContext, useContext, useRef } from 'react';
import { useFocusable } from 'react-aria/useFocusable';
import { TooltipTrigger as TooltipTrigger$1, OverlayArrow, Tooltip } from 'react-aria-components/Tooltip';
import { useCSSVariable } from '../../hooks/use-css-variable.js';
import { composeTwRenderProps, composeSlotClassName } from '../../utils/compose.js';
import { parseCSSTime } from '../../utils/css.js';
import { dom } from '../../utils/dom.js';
import { jsx } from 'react/jsx-runtime';

const TooltipContext = /*#__PURE__*/createContext({});

/* -------------------------------------------------------------------------------------------------
 * Tooltip Root
 * -----------------------------------------------------------------------------------------------*/

const TooltipRoot = ({
  children,
  closeDelay,
  delay,
  ...props
}) => {
  const slots = React__default.useMemo(() => tooltipVariants(), []);
  const cssDelay = useCSSVariable("--tooltip-delay");
  const cssCloseDelay = useCSSVariable("--tooltip-close-delay");
  const resolvedDelay = delay ?? parseCSSTime(cssDelay);
  const resolvedCloseDelay = closeDelay ?? parseCSSTime(cssCloseDelay);
  return /*#__PURE__*/jsx(TooltipContext, {
    value: {
      slots
    },
    children: /*#__PURE__*/jsx(TooltipTrigger$1, {
      closeDelay: resolvedCloseDelay,
      "data-slot": "tooltip-root",
      delay: resolvedDelay,
      ...props,
      children: children
    })
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tooltip Content
 * -----------------------------------------------------------------------------------------------*/

const TooltipContent = ({
  children,
  className,
  offset: offsetProp,
  showArrow = false,
  ...props
}) => {
  const {
    slots
  } = useContext(TooltipContext);
  const offset = offsetProp ? offsetProp : showArrow ? 7 : 3;
  return /*#__PURE__*/jsx(Tooltip, {
    ...props,
    className: composeTwRenderProps(className, slots?.base()),
    offset: offset,
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tooltip Arrow
 * -----------------------------------------------------------------------------------------------*/

const TooltipArrow = ({
  children,
  className,
  ...props
}) => {
  const defaultArrow = /*#__PURE__*/jsx("svg", {
    "data-slot": "overlay-arrow",
    fill: "none",
    height: "12",
    viewBox: "0 0 12 12",
    width: "12",
    xmlns: "http://www.w3.org/2000/svg",
    children: /*#__PURE__*/jsx("path", {
      d: "M0 0C5.48483 8 6.5 8 12 0Z"
    })
  });
  const arrow = /*#__PURE__*/React__default.isValidElement(children) ? /*#__PURE__*/React__default.cloneElement(children, {
    "data-slot": "overlay-arrow"
  }) : defaultArrow;
  return /*#__PURE__*/jsx(OverlayArrow, {
    "data-slot": "tooltip-arrow",
    ...props,
    className: className,
    children: arrow
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tooltip Trigger
 * -----------------------------------------------------------------------------------------------*/

const TooltipTrigger = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = useContext(TooltipContext);
  const triggerRef = useRef(null);
  // Use the `useFocusable` hook directly rather than the `<Focusable>` wrapper component.
  // Both consume the FocusableContext provided by TooltipTriggerPrimitive's internal
  // <FocusableProvider> (which carries the tooltip trigger props + triggerRef) and produce
  // the same `focusableProps` (tabIndex, focus, keyboard handlers). However, the wrapper
  // component runs a dev-only `isFocusable(el)` assertion in a useEffect that walks the
  // ancestor chain for `inert`. When the Tooltip mounts inside an inert subtree (e.g. behind
  // an open Drawer/Modal), that check produces a false-positive warning:
  // "<Focusable> child must be focusable. Please ensure the tabIndex prop is passed through."
  const {
    focusableProps
  } = useFocusable({}, triggerRef);
  return /*#__PURE__*/jsx(dom.div, {
    ref: triggerRef,
    className: composeSlotClassName(slots?.trigger, className),
    "data-slot": "tooltip-trigger",
    role: "button",
    ...mergeProps(focusableProps, props),
    children: children
  });
};

export { TooltipArrow, TooltipContent, TooltipRoot, TooltipTrigger };
